android/src/main/java/com/tailscale/ipn/ui/localapi/Client.kt

```diff
@@ -0,0 +1,230 @@
+        ).execute()
+    }
+
+    private inline fun <reified T> post(
+        path: String, body: ByteArray? = null, noinline responseHandler: (Result<T>) -> Unit
+    ) {
+        Request(
+            scope = scope,
+            method = "POST",
+            path = path,
+            body = body,
+            responseType = typeOf<T>(),
+            responseHandler = responseHandler
+        ).execute()
+    }
+
+    private inline fun <reified T> patch(
+        path: String, body: ByteArray? = null, noinline responseHandler: (Result<T>) -> Unit
+    ) {
+        Request(
+            scope = scope,
+            method = "PATCH",
+            path = path,
+            body = body,
+            responseType = typeOf<T>(),
+            responseHandler = responseHandler
+        ).execute()
+    }
+
+    private inline fun <reified T> delete(
+        path: String, noinline responseHandler: (Result<T>) -> Unit
+    ) {
+        Request(
+            scope = scope,
+            method = "DELETE",
+            path = path,
+            responseType = typeOf<T>(),
+            responseHandler = responseHandler
+        ).execute()
+    }
+}
+
+class Request<T>(
+    private val scope: CoroutineScope,
+    private val method: String,
+    path: String,
+    private val body: ByteArray? = null,
+    private val responseType: KType,
+    private val responseHandler: (Result<T>) -> Unit
+) {
+    private val fullPath = "/localapi/v0/$path"
+
+    companion object {
+        private const val TAG = "LocalAPIRequest"
+
+        private val jsonDecoder = Json { ignoreUnknownKeys = true }
+        private val isReady = CompletableDeferred<Boolean>()
+
+        // Called by the backend when the localAPI is ready to accept requests.
+        @JvmStatic
+        @Suppress("unused")
+        fun onReady() {
+            isReady.complete(true)
+            Log.d(TAG, "Ready")
+        }
+    }
+
+    // Perform a request to the local API in the go backend.  This is
+    // the primary JNI method for servicing a localAPI call. This
+    // is GUARANTEED to call back into onResponse.
+    // @see cmd/localapiclient/localapishim.go
+    //
+    // method: The HTTP method to use.
+    // request: The path to the localAPI endpoint.
+    // body: The body of the request.
+    private external fun doRequest(method: String, request: String, body: ByteArray?)
+
+    fun execute() {
+        scope.launch(Dispatchers.IO) {
+            isReady.await()
+            Log.d(TAG, "Executing request:${method}:${fullPath}")
+            doRequest(method, fullPath, body)
+        }
+    }
+
+    // This is called from the JNI layer to publish responses.
+    @OptIn(ExperimentalSerializationApi::class)
+    @Suppress("unused", "UNCHECKED_CAST")
+    fun onResponse(respData: ByteArray) {
+        Log.d(TAG, "Response for request: $fullPath")
+
+        val response: Result<T> = when (responseType) {
+            typeOf<String>() -> Result.success(respData.decodeToString() as T)
```

Comment:
Instead of each endpoint's response handler specifying its own decoder, we now just decode appropriately based on the datatype that the response handler requires.